Feature/bench

Davidson-Souza opened the pull request:
Add some simple benchmarks to this lib. Benchmarks require a nightly channel to run because they are unstable features. You can run with `RUSTFLAGS='--cfg=bench' cargo +nightly test`

kcalvinalvin reviewed (commented):
The benchmarks look ok for the most part to me.

Two things though:

Getting this error:

```
[I] calvin@Calvins-MacBook-Pro ~/b/r/rustreexo ((0a218725…))> RUSTFLAGS='--cfg=bench' cargo +nightly test
   Compiling serde_json v1.0.81                                    
   Compiling rustreexo v0.1.0 (/Users/calvin/bitcoin-projects/rustreexo/rustreexo)
error[E0432]: unresolved import `crate::accumulator::util::hash_from_u8`
   --> src/accumulator/types.rs:37:9                      
    |                                                              
37  |     use crate::accumulator::util::hash_from_u8;     
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `hash_from_u8` in `accumulator::util`
    |                                                           
note: found an item that was configured out              
   --> src/accumulator/util.rs:374:15                                                                                                  
    |                                                                                                                                  
374 | pub(crate) fn hash_from_u8(value: u8) -> sha256::Hash {                                                                          
    |               ^^^^^^^^^^^^                 
```

It was resolved for me by adding the `test` attribute.

```diff
diff --git a/src/accumulator/types.rs b/src/accumulator/types.rs
index 0e3faa0..7593d64 100644
--- a/src/accumulator/types.rs
+++ b/src/accumulator/types.rs
@@ -33,6 +33,7 @@ mod test {
     }
 }
 #[cfg(bench)]
+#[cfg(test)]
 mod bench {
     use crate::accumulator::util::hash_from_u8;
     extern crate test;
```

I don't think the actual benchmarks run unless it's `RUSTFLAGS='--cfg=bench' cargo +nightly bench`.

Davidson-Souza commented on src/accumulator/stump.rs:
```diff
@@ -445,3 +445,56 @@ mod test {
+        });
+    }
+    #[bench]
+    fn bench_add_del(bencher: &mut Bencher) {
```
Because it only tests against one case? I can add more values if needed.

kcalvinalvin replied:
> Because it only tests against one case?

Yeah because of that.


Also when full pollard is working, it might be good to add to a really high tree and delete there? Since for Bitcoin use-cases, the forest will be tall (like 2^28). Probably not the best to add that much for the sake of the benchmark running quick enough but maybe like 10,000 leaves definitely seem good.

It's definitely not easy to do that since you need to hard code the proof in but if full pollard is working, you don't need to do that anymore and can just generate the proof from the pollard.

Davidson-Souza commented:
Forced-push 7e2074d, rebased with `main` and addresses the review suggestions.

> 1: Getting this error:

See if it works now.

> 2: I don't think the actual benchmarks run unless it's RUSTFLAGS='--cfg=bench' cargo +nightly bench.

This is true, but currently it can't be fixed. Benchmarks use the unstable feature `test` in rust, so you need to use the +nightly. If we don't hide the bench behind a `#[cfg(bench)]` (a feature flag that you enable with the `RUSTFLAGS='--cfg=bench'`), any non-nightly build would break.

Davidson-Souza commented on src/accumulator/stump.rs:
```diff
@@ -445,3 +445,56 @@ mod test {
+        )
+        .unwrap();
+        bencher.iter(|| {
+            let _ = Stump::new().modify(&vec![hash], &vec![], &Proof::default());
```
You're right, there's no need to use it here, since we only borrow anyway. I moved both constructions to outside the bench closure.

Davidson-Souza commented on src/accumulator/types.rs:
```diff
@@ -39,3 +32,17 @@ mod test {
+    use test::Bencher;
+    #[bench]
+    fn bench_parent_hash(bencher: &mut Bencher) {
+        let hash1 = hash_from_u8(0);
```
No, only what goes inside the `bencher.iter` closure.

kcalvinalvin commented:
ACK 7e2074d7131bec948d2dfced22b714d88ec1d3e3

For https://github.com/mit-dci/rustreexo/pull/28#discussion_r1260005694, we could just merge this and add extras as another PR. Up to you really.

Davidson-Souza commented:
> For https://github.com/mit-dci/rustreexo/pull/28#discussion_r1260005694, we could just merge this and add extras as another PR. Up to you really.

I'm inclined towards that. I'll make micro benchmarks for internal code in a follow-up too.
